### PR TITLE
chore: add grpc based auth fallback to proxied requests

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

We have multiple middlewares that cover authentication. The master proxy we use for GAS does not engage the one that
checks for allocation tokens.

## Test Plan

added integ tests

## Commentary (optional)

the logic checking for allocation tokens should probably sit somewhere as a global middleware that encompasses user.ProcessAuth and allocations 🤔 

https://github.com/determined-ai/determined/blob/6ffb2f64513619ffe8fbab23529b1d71e931e6ef/master/internal/grpcutil/api.go#L69

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

blocks https://hpe-aiatscale.atlassian.net/browse/GAS-507?filter=-1

### CODE REVIEW

File: master/internal/api_auth.go [hunk 1]
```diff
@@ -168,19 +169,44 @@ func processAuthWithRedirect(redirectPaths []string) echo.MiddlewareFunc {
...
+				}
+				_, _, err = grpcutil.GetUser(metadata.NewIncomingContext(c.Request().Context(), md))
+				if err == nil {
+					return next(c)
+				}
+			} else {
+				return err
+			}
+
+			// TODO: revert this logic to redirect only if accept is empty or specifies text/html.
```
Comment (hamidzr) [inline master/internal/api_auth.go:201]: reverse. maybe separate pr?

File: master/internal/api_user_intg_test.go [hunk 2]
```diff
@@ -172,54 +173,116 @@ func TestProcessAuth(t *testing.T) {
...
+		AllocationID: model.AllocationID(fmt.Sprintf("%s-1", tIn.TaskID)),
+		TaskID:       tIn.TaskID,
+		StartTime:    ptrs.Ptr(time.Now().UTC().Truncate(time.Millisecond)),
+		State:        ptrs.Ptr(model.AllocationStateTerminated),
+	}
+
+	err := db.AddAllocation(ctx, &a)
+	require.NoError(t, err, "failed to add allocation")
+
+	res, err := db.AllocationByID(ctx, a.AllocationID)
```
Comment (salonig23) [inline master/internal/api_user_intg_test.go:190]: assert a equals value of res

File: master/internal/api_user_intg_test.go [hunk 3]
```diff
@@ -172,54 +173,116 @@ func TestProcessAuth(t *testing.T) {
...
+		User: &userv1.User{
+			Username: username,
+			Active:   true,
+		},
+		Password: "testpassword",
+	})
+	require.NoError(t, err)
+
+	user := model.User{Username: username, ID: model.UserID(resp.User.Id)}
+	require.NoError(t, err)
```
Comment (salonig23) [inline master/internal/api_user_intg_test.go:212]: Unnecessary check

File: master/internal/api_user_intg_test.go [hunk 4]
```diff
@@ -172,54 +173,116 @@ func TestProcessAuth(t *testing.T) {
...
+
+	allocation := setupNewAllocation(t, api.m.db)
+	allocationToken, err := db.StartAllocationSession(ctx, allocation.AllocationID, &user)
```
Comment (salonig23) [inline master/internal/api_user_intg_test.go:215]: assert allocationToken is not empty

**Review by salonig23 — APPROVED** [review]
LGTM apart from minor comments

**Pull request merged**